### ISSUE TITLE
False positive with unreachable code in operator overload

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#9595]
```py
class A:
    def __and__(self, other: int) -> str:
        if isinstance(other, int):
            return "foo"
        return NotImplemented

class B:
    def __rand__(self, other: A) -> str:
        return "bar"

print(A() & B())
```
```
error: Statement is unreachable
            return NotImplemented
            ^
```

**Comment by hauntsaninja** [comment]
Your type annotation is incorrect, use `other: object`

**Comment by KotlinIsland** [comment]
@hauntsaninja That introduces false negatives though.
```py
class A:
    def __and__(self, other: object) -> str:
        if isinstance(other, int):
            return "foo"
        return NotImplemented

class B:
    def __rand__(self, other: A) -> str:
        return "bar"

print(A() & B())
print(A() & "asdf")  // no mypy error but TypeError: unsupported operand type(s) for &: 'A' and 'str'
```

**Comment by hauntsaninja** [comment]
Use a protocol that defines rand then

**Comment by KotlinIsland** [comment]
Here is how to implement it for anyone that finds this:
```py
from __future__ import annotations

from typing import Protocol, TypeVar, Union

T = TypeVar("T")
R = TypeVar("R")


class Rand(Protocol[T, R]):
    def __rand__(self, other: T) -> R:
        ...


class A:
    def __and__(self, other: Union[int, Rand[A, str]]) -> str:
        if isinstance(other, int):
            return "foo"
        return NotImplemented


class B:
    def __rand__(self, other: A) -> str:
        return "bar"


print(A() & B())
print(A() & "asdf")
# error: Unsupported operand types for & ("A" and "str")
#     print(A() & "asdf")
#                 ^
```
I think this should be implicit default behavior of mypy though.

**Comment by JelleZijlstra** [comment]
It's indeed unfortunate if something like the `__rand__` protocol was required to correctly type a dunder method with mypy. The way the code was written in the original post is much more clear and concise.

We already have a special case that allows returning `NotImplemented` in dunders (`is_literal_not_implemented` in `mypy/checker.py`), so it seems reasonable to also add a special case for the warning the OP reported here. Unfortunately, it's probably not easy to implement.
